datalad/distribution/update.py

```diff
@@ -109,82 +104,67 @@ def __call__(
-            # what we want? Do we want to specify a refspec instead?
-            # yoh: we should leave it to git and its configuration.
-            # So imho we should just extract to fetch everything git would fetch
-            if knows_annex(repo.path) and not fetch_all:
-                if remote:
-                    # we are updating from a certain remote, so git-annex branch
-                    # should be updated from there as well:
-                    repo.fetch(remote=remote)
-                    # TODO: what does failing here look like?
-                else:
-                    # we have no remote given, therefore
-                    # check for tracking branch's remote:
-
-                    track_remote, track_branch = repo.get_tracking_branch()
-                    if track_remote:
-                        # we have a "tracking remote"
-                        repo.fetch(remote=track_remote)
-
-            # merge:
-            if merge:
-                lgr.info("Applying changes from tracking branch...")
-                # TODO: Adapt.
-                # TODO: Rethink default remote/tracking branch. See above.
-                # We need a "tracking remote" but custom refspec to fetch from
-                # that remote
-                cmd_list = ["git", "pull"]
-                if remote:
-                    cmd_list.append(remote)
-                    # branch needed, if not default remote
-                    # => TODO: use default remote/tracking branch to compare
-                    #          (see above, where git-annex is fetched)
-                    # => TODO: allow for passing a branch
-                    # (or more general refspec?)
-                    # For now, just use the same name
-                    cmd_list.append(repo.get_active_branch())
-
-                std_out, std_err = repo._git_custom_command('', cmd_list)
-                lgr.info(std_out)
-                if knows_annex(repo.path):
-                    # annex-apply:
-                    lgr.info("Updating annex ...")
-                    std_out, std_err = repo._git_custom_command(
-                        '', ["git", "annex", "merge"])
-                    lgr.info(std_out)
-
-                    # TODO: return value?
+            _update_repo(ds, sibling_, merge, fetch_all, reobtain_data)
+
+
+def _update_repo(ds, remote, merge, fetch_all, reobtain_data):
+    repo = ds.repo
+    # fetch remote
+    repo.fetch(
+        remote=None if fetch_all else remote,
+        all_=fetch_all,
+        prune=True)  # prune to not accumulate a mess over time
+
+    if not merge:
+        return
+    # we need to check whether we need to convert this dataset to
+    # annex, would would be the case when we presently have a git repo
+    # and the recent fetch brought evidence for a remote annex
+    if not isinstance(repo, AnnexRepo) and knows_annex(repo.path):
+        lgr.info("Init annex at '%s' prior merge.", repo.path)
```

Comment:
I think, this should be done be re-requesting `ds.repo`, since that property already has a check for whether it became an annex. This check in `Dataset.repo` should use `knows_annex` though, since it currently is just looking for a local git-annex' branch.